add support for variable routing plane count based on live ethernet link count

SeanNijjar opened the pull request:
### Problem description
Current implementation of fabric initiailization does not reliably handle systems where ethernet links are unreliable/disconnected.

### What's changed
On control plane/fabric initialization, support is added to instantiate only as many routing planes as there are live links available. The general policy is as follows:
1) For 1D topologies, rows/columns are evaluated independently
  - The full row/column of devices is traversed and the narrowest step of the traversal is applied as the number of routing planes
  - For example, if there are 4 hops, and the link counts per hop are {2, 3, 4, 4}, then 2 routing planes will be instantiated for this 1D topology. Different rows/columns of the device will have different routing planes depending on live link counts
2) For 2D topologies, the minimal number of routing planes across all hops is applied to the full fabric.

The implementation for reducing the number of routing planes is to trim the number of active ethernet links as a post process step after discovering all links. This is done before routing tables are generated. As a more robust implementation would have a separate datastructure (or set of fields) to differentiate between routing plane ethernet cores and active ethernet cores. With that approach, control plane could understand remaining cores available if it wants to reprogram routing planes during long running workloads

- Build reference cluster connectivity (same as before)
- Compute number of routing planes per row and column (for 1D topologies) or for all rows/cols collectively (for mesh).
- Trim active ethernet cores according to the maximum number of routing planes, for full rows and columns

## Visual Example

Consider a healthy 4-chip line topology with 4 routing planes:

```                                                                                                                           
 ┌───────────────────┐             ┌───────────────────┐             ┌───────────────────┐             ┌───────────────────┐ 
 │                   │    RP 0     │                   │    RP 0     │                   │    RP 0     │                   │ 
 │                   │=============│                   │=============│                   │=============│                   │ 
 │                   │             │                   │             │                   │             │                   │ 
 │                   │             │                   │             │                   │             │                   │ 
 │                   │    RP 1     │                   │    RP 1     │                   │    RP 1     │                   │ 
 │                   │=============│                   │=============│                   │=============│                   │ 
 │                   │             │                   │             │                   │             │                   │ 
 │                   │             │                   │             │                   │             │                   │ 
 │                   │    RP 2     │                   │    RP 2     │                   │    RP 2     │                   │ 
 │                   │=============│                   │=============│                   │=============│                   │ 
 │                   │             │                   │             │                   │             │                   │ 
 │                   │             │                   │             │                   │             │                   │ 
 │                   │    RP 3     │                   │    RP 3     │                   │    RP 3     │                   │ 
 │                   │=============│                   │=============│                   │=============│                   │ 
 │                   │             │                   │             │                   │             │                   │ 
 │                   │             │                   │             │                   │             │                   │ 
 └───────────────────┘             └───────────────────┘             └───────────────────┘             └───────────────────┘ 
```

The (chip) internal router-to-router connections would look something like this:
```
         Chip 0                           Chip 1                           Chip 2                             Chip 3          
  ┌───────────────────┐             ┌───────────────────┐             ┌───────────────────┐             ┌───────────────────┐ 
  │                   │    RP 0     │                   │    RP 0     │                   │    RP 0     │                   │ 
  │                   │=============│<----------------->│=============│<----------------->│=============│                   │ 
  │                   │             │                   │             │                   │             │                   │ 
  │                   │             │                   │             │                   │             │                   │ 
  │                   │    RP 1     │                   │    RP 1     │                   │    RP 1     │                   │ 
  │                   │=============│<----------------->│=============│<----------------->│=============│                   │ 
  │                   │             │                   │             │                   │             │                   │ 
  │                   │             │                   │             │                   │             │                   │ 
  │                   │    RP 2     │                   │    RP 2     │                   │    RP 2     │                   │ 
  │                   │=============│<----------------->│=============│<----------------->│=============│                   │ 
  │                   │             │                   │             │                   │             │                   │ 
  │                   │             │                   │             │                   │             │                   │ 
  │                   │    RP 3     │                   │    RP 3     │                   │    RP 3     │                   │ 
  │                   │=============│<----------------->│=============│<----------------->│=============│                   │ 
  │                   │             │                   │             │                   │             │                   │ 
  │                   │             │                   │             │                   │             │                   │ 
  └───────────────────┘             └───────────────────┘             └───────────────────┘             └───────────────────┘
```

However, for an unhealthy system, some links may not be active or live. In those cases, the control plane must build fabric differently. In this example, consider the first link dead for the first edge and the last link dead for the last edge. In that scenario, the routing planes could be configured as shown. One of the edges still has 4 live links, but only 3 would participate in that fabric as routing planes.

```
         Chip 0                           Chip 1                           Chip 2                             Chip 3          
  ┌───────────────────┐             ┌───────────────────┐             ┌───────────────────┐             ┌───────────────────┐ 
  │                   │             │                   │   RP 0      │                   │   RP 0      │                   │ 
  │                   │==X       X==│           /------>│=============│<----------------->│=============│                   │ 
  │                   │             │          /        │             │                   │             │                   │ 
  │                   │             │         /         │             │                   │             │                   │ 
  │                   │   RP 0      │        /          │   RP 1      │                   │   RP 1      │                   │ 
  │                   │=============│<------/   /------>│=============│<----------------->│=============│                   │ 
  │                   │             │          /        │             │                   │             │                   │ 
  │                   │             │         /         │             │                   │             │                   │ 
  │                   │   RP 1      │        /          │   RP 2      │                   │   RP 2      │                   │ 
  │                   │=============│<------/   /------>│=============│<----------------->│=============│                   │ 
  │                   │             │          /        │             │                   │             │                   │ 
  │                   │             │         /         │             │                   │             │                   │ 
  │                   │   RP 2      │        /          │  INACTIVE   │                   │             │                   │ 
  │                   │=============│<------/           │=============│                   │==X       X==│                   │ 
  │                   │             │                   │             │                   │             │                   │ 
  │                   │             │                   │             │                   │             │                   │ 
  └───────────────────┘             └───────────────────┘             └───────────────────┘             └───────────────────┘ 
```


## User Facing APIs
Added a new user facing kernel to determine the number of live routing planes, given a cluster axis and direction in the mesh of devices:

`tt::tt_fabric::experimental::get_number_of_available_routing_planes()`

An example of how this API may be used at op level for CCLs:

```
    std::optional<size_t> num_links = this->num_links_preferred;

    if (!num_links.has_value()) {
        if (this->cluster_axis.has_value()) {
            auto row_or_col = this->cluster_axis.value() == 0 ? coord[1] : coord[0];
            num_links = tt::tt_fabric::experimental::get_number_of_available_routing_planes(
                *mesh_device, this->cluster_axis.value(), row_or_col);
        } else {
            TT_FATAL(
                false,
                "Reduce scatter with undefined number of links and non-cluster axis API is currently not supported");
        }
    }
``` 

## Future Work
- Add more extensive testing where a test harness manually disables links and runs fabric tests
- Add downshifting from ring down to mesh topologies in cases where all links are down between pairs of chips

### Ticket
https://github.com/tenstorrent/tt-metal/issues/23189

### Checklist
- [x] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI: https://github.com/tenstorrent/tt-metal/actions/runs/15716594434
- [x] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes: https://github.com/tenstorrent/tt-metal/actions/runs/15721262254
- [x] T3K: https://github.com/tenstorrent/tt-metal/actions/runs/15716598314
- [x] TG: https://github.com/tenstorrent/tt-metal/actions/runs/15716603009
- [x] New/Existing tests provide coverage for changes

SeanNijjar commented on tt_metal/api/tt-metalium/control_plane.hpp:
```diff
@@ -67,8 +79,16 @@ class ControlPlane {
     std::vector<chan_id_t> get_active_fabric_eth_channels_in_direction(
         FabricNodeId fabric_node_id, RoutingDirection routing_direction) const;
+    // Return the active routing planes in a given direction.
+    std::vector<chan_id_t> get_active_fabric_eth_routing_planes_in_direction(
+        FabricNodeId fabric_node_id, RoutingDirection routing_direction) const;
+
+    size_t get_num_available_routing_planes_in_direction(
```
Agreed in the sense that a routing plane's scope exceeds the scope of a single fabric node. However this let's us query at the fabric node ID, which I found was more usable for other places in control plane.

However,if you look at the fabric APIs, you'll see that we expose the API to take  cluster_axis + row/col instead of node ID. Does this answer your question?

aagarwalTT commented on tt_metal/fabric/control_plane.cpp:
```diff
@@ -86,6 +86,234 @@ bool is_chip_on_corner_of_mesh(
+                // here, val is initialized to max
+                if (reliability_mode == tt_metal::FabricReliabilityMode::STRICT_SYSTEM_HEALTH_SETUP_MODE) {
+                    TT_FATAL(
+                        golden_link_counts.at(direction) == it->second.size(),
```
Can you please elaborate a bit on when the golden counts won't match here? It somewhat seems like both are derived from the same source?

SeanNijjar replied:
Yes, I think you are right. This check looks redundant.

SeanNijjar commented on tt_metal/fabric/control_plane.cpp:
```diff
@@ -86,6 +86,234 @@ bool is_chip_on_corner_of_mesh(
+        for (std::uint32_t chip_id = 0; chip_id < intra_mesh_connectivity[mesh_id].size(); chip_id++) {
+            const auto fabric_node_id = FabricNodeId(MeshId{mesh_id}, chip_id);
+            const auto chip_coord = get_chip_coord(fabric_node_id);
+            auto chip_coord_x = chip_coord[0];
```
Do you mean changing the iteration order to avoid the repeated lookup in `get_chip_coord`? Yes, that can be done when I clean this PR after I've received feedback for all functional/design concerns.

aagarwalTT replied:
I guess we will still need some idea of rows/columns as we can return different number of active routing planes per row/column and only direction might not be sufficient here

yugaoTT commented on tt_metal/fabric/control_plane.cpp:
```diff
@@ -1037,6 +1264,21 @@ std::vector<chan_id_t> ControlPlane::get_active_fabric_eth_channels_in_direction
     return {};
 }
 
+std::vector<chan_id_t> ControlPlane::get_active_fabric_eth_routing_planes_in_direction(
+    FabricNodeId fabric_node_id, RoutingDirection routing_direction) const {
+    auto eth_chans = get_active_fabric_eth_channels_in_direction(fabric_node_id, routing_direction);
+    auto num_routing_planes = get_num_available_routing_planes_in_direction(fabric_node_id, routing_direction);
+    TT_FATAL(eth_chans.size() >= num_routing_planes, "Not enough active fabric eth channels in direction");
+    eth_chans.resize(num_routing_planes);
+    return eth_chans;
+}
+
+size_t ControlPlane::get_num_available_routing_planes_in_direction(
+    FabricNodeId fabric_node_id, RoutingDirection routing_direction) const {
+    const auto& eth_chans = get_active_fabric_eth_channels_in_direction(fabric_node_id, routing_direction);
+    return eth_chans.size();
+}
```
in both functions they are calling the same get_active_fabric_eth_channels_in_direction?

aliuTT reviewed (approved):
We also want to allow fabric init to relax Torus to Mesh, probably can add an enum option here: `FabricReliabilityMode`

Should we have the Relaxed Mode as default for `set_fabric_config`?

SeanNijjar commented:
> We also want to allow fabric init to relax Torus to Mesh, probably can add an enum option here: `FabricReliabilityMode`
> 
> Should we have the Relaxed Mode as default for `set_fabric_config`?

Yes, this is future PR/work. I think in general, it would make sense to have the default mode be downshifted/relaxed, but I'd prefer to make that change in a separate PR. Strict mode should IMO be an opt-in.

SeanNijjar committed:
WIP dynamic routing plane count

WIP - support for dynamic number of links

WIP: add query API for CCLs for dynamic number of routing planes per CCL (row,column)

flexible link count test (DON'T MERGE)

undo accidental deletion

fixes after rebase

checkpoint

it works

bit fo cleanup

a little more cleanup

more cleanup

flip args

cleanup

cleanup

cleanup

move new API to experimental

fix some test regressions

cleanup

delete strict check for now. It's not robust enough yet.

special case handling for custom

delete check that's not robust enough

update metal context fabric config when setting fabric context config

sq

sq

ignore non-user-visible meshes (TG) when setting up relaxed mode routing planes

skip single chip -- no intra mesh connections

don't trim in non-user facing meshes (special case for TG)

better routing plane count DS init

refine

finally fixed?

yikes... embarrassing

cleanup

cleanup

restore && cleanup

more cleanup and fix an issue

aagarwalTT commented on tt_metal/fabric/control_plane.cpp:
```diff
@@ -1164,9 +1422,10 @@ void ControlPlane::set_routing_mode(uint16_t mode) {
+void ControlPlane::initialize_fabric_context(tt_metal::FabricConfig fabric_config, tt_metal::FabricReliabilityMode reliability_mode) {
     TT_FATAL(this->fabric_context_ == nullptr, "Trying to re-initialize fabric context");
-    this->fabric_context_ = std::make_unique<FabricContext>(fabric_config);
+    tt::tt_metal::MetalContext::instance().set_fabric_config(fabric_config, reliability_mode);
```
I don't think this will have any effect? In the current flow we reach here only if set_fabric_config was called before this step.

SeanNijjar replied:
This happens in some unit tests. I think as a worst case, this would force thye metal_context and control_plane to be in agreement. That being said, the fact that this information exists in two locations is a problem and should be remedied in the future.

aagarwalTT replied:
Understood, just wanted to point out that this would be mostly a no-op other than updating the `fabric_config_` in metal context

aagarwalTT commented on tt_metal/fabric/control_plane.cpp:
```diff
@@ -1164,9 +1422,10 @@ void ControlPlane::set_routing_mode(uint16_t mode) {
     TT_FATAL(this->fabric_context_ == nullptr, "Trying to re-initialize fabric context");
-    this->fabric_context_ = std::make_unique<FabricContext>(fabric_config);
+    tt::tt_metal::MetalContext::instance().set_fabric_config(fabric_config, reliability_mode);
+    this->fabric_context_ = std::make_unique<FabricContext>(fabric_config);//, reliability_mode);
```
remove comment?

aagarwalTT commented on tt_metal/fabric/fabric.cpp:
```diff
@@ -12,11 +12,14 @@
 #include <set>
 #include <vector>
 
+#include "distributed/mesh_socket_utils.hpp"
```
why do we need the socket utils header here?

SeanNijjar replied:
it's for `distributed::get_physical_mesh_id`

aagarwalTT replied:
ahh I see. Just a thought- while calculating the dynamic routing plane counts, we do maintain counts per row and columns. Can we just store them and query them in host side API so that  we don't have to do any conversion from row/col to a device?

SeanNijjar replied:
Yes I definitely think we can explore more efficient lookup but I think we should get the APIS and control plane stack sorted before deciding what to cache and how.

aagarwalTT commented on tt_metal/fabric/fabric_context.hpp:
```diff
@@ -25,7 +25,10 @@ class FabricContext {
+    static tt::tt_fabric::Topology get_topology_from_config(tt::tt_metal::FabricConfig fabric_config);
+
     tt::tt_fabric::Topology get_fabric_topology() const;
+    tt::tt_metal::FabricConfig get_fabric_config() const { return fabric_config_; }
```
remove? we already have a getter in metal context for this

aagarwalTT replied:
yeah but that's only cached for local init/setup within the fabric context

SeanNijjar replied:
I see. Are they expected to differ after initial setup?

SeanNijjar committed:
cleanup

cfjchu commented on tt_metal/api/tt-metalium/control_plane.hpp:
```diff
@@ -107,6 +122,9 @@ class ControlPlane {
     // If the ethernet core is an intermesh link, probe to see if it is trained
     bool is_intermesh_eth_link_trained(chip_id_t chip_id, CoreCoord eth_core) const;
 
+    // Calculate number of active routing planes for each mesh
+    size_t get_num_active_routing_planes(const IntraMeshConnectivity& intra_mesh_connectivity);
```
I don't see an implementation for this. Can this be removed?

cfjchu commented on conftest.py:
```diff
@@ -413,6 +413,7 @@ def mesh_device(request, silicon_arch_name, device_params):
 
     updated_device_params = get_updated_device_params(device_params)
     fabric_config = updated_device_params.pop("fabric_config", None)
+    reliability_mode = updated_device_params.pop("reliability_mode", None)
```
There's a number of spots in this conftest.py where we do:

```
fabric_config = updated_device_params.pop("fabric_config", None)
```


do those other instances also need update?

cfjchu commented on tests/ttnn/unit_tests/gtests/ccl/test_fabric_edm_common.hpp:
```diff
@@ -148,13 +148,88 @@ class Fabric1DFixture : public BaseFabricFixture {
+
+    std::shared_ptr<MeshDevice> mesh_device_;
+
+    // Gets the appropriate mesh shape based on device configuration
+    MeshShape GetDeterminedMeshShape() const {
+        if (num_devices_ == TG_NUM_DEVICES || num_devices_ == GALAXY_6U_NUM_DEVICES) {
+            return MeshShape{8, 4};
+        } else {
+            return MeshShape{2, 4};
+        }
+    }
```
You don't need to do this. You can just call `SystemMesh::instance().get_shape`

cfjchu commented on tests/ttnn/unit_tests/gtests/ccl/test_fabric_edm_common.hpp:
```diff
@@ -3594,29 +3687,115 @@ void RunWriteThroughputStabilityTestWithPersistentFabric(
+    size_t num_links = num_links_opt.value_or(0);
+    if (num_links == 0) {
+        if (cluster_type == ClusterType::GALAXY) {
+            for (size_t i = 0; i < 8; i++) {
```
why hardcoded rows/columns?

SeanNijjar replied:
Please forgive my moment of laziness. Updated now to  `mesh_device.shape()[0]` and `mesh_device.shape()[1]` respectively.

SeanNijjar committed:
cleanup

SeanNijjar committed:
moar

SeanNijjar committed:
fix regression

SeanNijjar commented on tt_metal/api/tt-metalium/control_plane.hpp:
```diff
@@ -142,6 +163,10 @@ class ControlPlane {
 
     void load_physical_chip_mapping(
         const std::map<FabricNodeId, chip_id_t>& logical_mesh_chip_id_to_physical_chip_id_mapping);
+    size_t get_num_live_routing_planes(FabricNodeId fabric_node_id, RoutingDirection routing_direction) const;
```
sorry I somehow missed this before merging. I'll open a new PR to add this
